File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PageItemsFetcherBasedDataSource.java
package com.azure.android.core.util.paging;

import androidx.annotation.NonNull;
import androidx.paging.PageKeyedDataSource;

import java.util.List;

/**
 * PACKAGE PRIVATE.
 *
 * An implementation of {@link PageKeyedDataSource} that uses {@link PageItemsFetcher}
 * abstraction to retrieve the pages.
 *
 * @param <T> the type of items in page
 */
class PageItemsFetcherBasedDataSource<T> extends PageKeyedDataSource<String, T> {
    private final PageItemsFetcher<T> pageItemsFetcher;
    private final DataSourceNotification notification;

    /**
     *  PACKAGE PRIVATE CTR.
     *
     *  Creates PageItemsFetcherBasedDataSource.
     *
     * @param pageItemsFetcher the fetcher to request pages
     * @param notification an object to notify various events in the DataSource
     */
    PageItemsFetcherBasedDataSource(PageItemsFetcher<T> pageItemsFetcher,
                                    DataSourceNotification notification) {
        this.pageItemsFetcher = pageItemsFetcher;
        this.notification = notification;
        this.notification.setInvalidateDataSourceRunnable(() -> {
            invalidate();
        });
    }

    @Override
    public void loadInitial(@NonNull LoadInitialParams<String> params,
                            @NonNull LoadInitialCallback<String, T> loadCallback) {
        this.notification.notifyLoadingState(true);
        this.pageItemsFetcher.fetchPage(null, params.requestedLoadSize, new PageItemsFetcher.FetchCallback<T>() {
            @Override
            public void onSuccess(List<T> items, String currentPageIdentifier, String nextPageIdentifier) {
                loadCallback.onResult(items, currentPageIdentifier, nextPageIdentifier);
                notification.notifyLoadedState(true);
            }

            @Override
            public void onFailure(Throwable throwable) {
                notification.notifyLoadErrorState(true, () -> loadInitial(params, loadCallback));
            }
        });
    }

    @Override
    public void loadBefore(@NonNull LoadParams<String> params,
                           @NonNull LoadCallback<String, T> callback) {
        // ignored
    }

    @Override
    public void loadAfter(@NonNull LoadParams<String> params,
                          @NonNull LoadCallback<String, T> loadCallback) {
        this.notification.notifyLoadingState(false);
        this.pageItemsFetcher.fetchPage(params.key, params.requestedLoadSize, new PageItemsFetcher.FetchCallback<T>() {
            @Override
            public void onSuccess(List<T> items, String currentPageIdentifier, String nextPageIdentifier) {
                loadCallback.onResult(items, nextPageIdentifier);
                notification.notifyLoadedState(false);
            }

            @Override
            public void onFailure(Throwable throwable) {
                notification.notifyLoadErrorState(false, () -> loadAfter(params, loadCallback));
            }
        });
    }
}

File: samples/sample-app-storage-kotlin/src/main/java/com/azure/android/storage/sample/kotlin/core/util/paging/PageItemsFetcherBasedDataSource.kt
package com.azure.android.storage.sample.kotlin.core.util.paging

import androidx.paging.PageKeyedDataSource

/**
 * PACKAGE PRIVATE.
 *
 * An implementation of [PageKeyedDataSource] that uses [PageItemsFetcher]
 * abstraction to retrieve the pages.
 *
 * @param T the type of items in page
 */
internal class PageItemsFetcherBasedDataSource<T>(pageItemsFetcher: PageItemsFetcher<T>,
                                                  notification: DataSourceNotification) : PageKeyedDataSource<String?, T>() {
    private val pageItemsFetcher: PageItemsFetcher<T> = pageItemsFetcher
    private val notification: DataSourceNotification = notification

    override fun loadInitial(params: LoadInitialParams<String?>,
                             loadCallback: LoadInitialCallback<String?, T>) {
        notification.notifyLoadingState(true)

        val fetchCallback = object : PageItemsFetcher.FetchCallback<T> {
            override fun onSuccess(items: List<T>, currentPageIdentifier: String?, nextPageIdentifier: String?) {
                loadCallback.onResult(items, currentPageIdentifier, nextPageIdentifier)
                notification.notifyLoadedState(true)
            }

            override fun onFailure(throwable: Throwable?) {
                notification.notifyLoadErrorState(true, Runnable { loadInitial(params, loadCallback) })
            }
        }
        pageItemsFetcher.fetchPage(null, params.requestedLoadSize, fetchCallback)
    }

    override fun loadBefore(params: LoadParams<String?>,
                            callback: LoadCallback<String?, T>) {
        // ignored
    }

    override fun loadAfter(params: LoadParams<String?>,
                           loadCallback: LoadCallback<String?, T>) {
        notification.notifyLoadingState(false)
        val fetchCallback = object : PageItemsFetcher.FetchCallback<T> {
            override fun onSuccess(items: List<T>, currentPageIdentifier: String?, nextPageIdentifier: String?) {
                loadCallback.onResult(items, nextPageIdentifier)
                notification.notifyLoadedState(false)
            }

            override fun onFailure(throwable: Throwable?) {
                notification.notifyLoadErrorState(false, Runnable { loadAfter(params, loadCallback) })
            }
        }
        pageItemsFetcher.fetchPage(params.key, params.requestedLoadSize, fetchCallback)
    }

    init {
        this.notification.setInvalidateDataSourceRunnable(Runnable { invalidate() })
    }
}
